homeassistant/components/mqtt/__init__.py

```diff
@@ -724,53 +610,101 @@ class Subscription:
-        will_message: Optional[Message],
-        birth_message: Optional[Message],
-        tls_version: Optional[int],
-    ) -> None:
+    def __init__(self, hass: HomeAssistantType, config_entry, conf,) -> None:
         """Initialize Home Assistant MQTT client."""
-        # We don't import them on the top because some integrations
+        # We don't import on the top because some integrations
         # should be able to optionally rely on MQTT.
-        # pylint: disable=import-outside-toplevel
-        import paho.mqtt.client as mqtt
+        import paho.mqtt.client as mqtt  # pylint: disable=import-outside-toplevel
 
         self.hass = hass
-        self.broker = broker
-        self.port = port
-        self.keepalive = keepalive
+        self.config_entry = config_entry
+        self.conf = conf
         self.subscriptions: List[Subscription] = []
-        self.birth_message = birth_message
         self.connected = False
         self._mqttc: mqtt.Client = None
         self._paho_lock = asyncio.Lock()
 
-        if protocol == PROTOCOL_31:
+        self.init_client()
+        self.config_entry.add_update_listener(self.async_config_entry_updated)
+
+    @staticmethod
+    async def async_config_entry_updated(hass, entry) -> None:
+        """Handle signals of config entry being updated.
+
+        This is a static method because a class method (bound method), can not be used with weak references.
+        Causes for this is config entry options changing.
+        """
+        self = hass.data[DATA_MQTT]
+
+        conf = hass.data.get(DATA_MQTT_CONFIG)
+        if conf is None:
+            conf = CONFIG_SCHEMA({DOMAIN: dict(entry.data)})[DOMAIN]
+
+        self.conf = _merge_config(entry, conf)
+        await self.async_disconnect()
+        self.init_client()
+        await self.async_connect()
+
+        await discovery.async_stop(hass)
+        if self.conf.get(CONF_DISCOVERY):
+            await _async_setup_discovery(hass, self.conf, entry)
+
+    def init_client(self):
+        """Initialize paho client."""
+        # We don't import on the top because some integrations
+        # should be able to optionally rely on MQTT.
+        import paho.mqtt.client as mqtt  # pylint: disable=import-outside-toplevel
+
+        if self.conf[CONF_PROTOCOL] == PROTOCOL_31:
             proto: int = mqtt.MQTTv31
         else:
             proto = mqtt.MQTTv311
 
+        client_id = self.conf.get(CONF_CLIENT_ID)
         if client_id is None:
             self._mqttc = mqtt.Client(protocol=proto)
         else:
             self._mqttc = mqtt.Client(client_id, protocol=proto)
 
+        username = self.conf.get(CONF_USERNAME)
+        password = self.conf.get(CONF_PASSWORD)
         if username is not None:
             self._mqttc.username_pw_set(username, password)
 
+        certificate = self.conf.get(CONF_CERTIFICATE)
+
+        # For cloudmqtt.com, secured connection, auto fill in certificate
+        if (
```

Comment:
I wonder if we can drop this and just always load the `certifi` certs. We shouldn't use `requests.certs.where()` but can actually call directly `import certifi; certifi.where()`